source/common/ssl/ssl_socket.h

```diff
@@ -67,29 +68,43 @@ class SslSocket : public Network::TransportSocket,
   mutable std::string cached_url_encoded_pem_encoded_peer_certificate_;
 };
 
-class ClientSslSocketFactory : public Network::TransportSocketFactory {
+class ClientSslSocketFactory : public Network::TransportSocketFactory,
+                               public Secret::SecretCallbacks,
+                               Logger::Loggable<Logger::Id::config> {
 public:
   ClientSslSocketFactory(ClientContextConfigPtr config, Ssl::ContextManager& manager,
                          Stats::Scope& stats_scope);
+  virtual ~ClientSslSocketFactory();
 
   Network::TransportSocketPtr createTransportSocket() const override;
   bool implementsSecureTransport() const override;
 
+  // Secret::SecretCallbacks
+  void onAddOrUpdateSecret() override;
+
 private:
   Ssl::ContextManager& manager_;
   Stats::Scope& stats_scope_;
   ClientContextConfigPtr config_;
   ClientContextSharedPtr ssl_ctx_;
 };
 
-class ServerSslSocketFactory : public Network::TransportSocketFactory {
+typedef std::unique_ptr<ClientContextConfig> ClientContextConfigPtr;
```

Comment:
I think ClientContextConfigPtr and ServerContextConfigPtr are already defined in include/envoy/ssl/context_config.h